internal/ansible/ansible.go

```diff
@@ -0,0 +1,221 @@
+		CRCDispatcherErrorCode     string `json:"crc_dispatcher_error_code"`
+		CRCDispatcherErrorDetails  string `json:"crc_dispatcher_error_details"`
+	} `json:"event_data"`
+}
+
+// RunPlaybook creates an ansible-runner job to run the provided playbook. The
+// function returns a channel over which the caller can receive ansible-runner
+// events as they happen. The channel is closed when the job completes.
+func RunPlaybook(id string, playbook []byte, correlationID string) (chan Event, error) {
+	privateDataDir := filepath.Join(constants.StateDir, id)
+
+	if err := os.WriteFile(filepath.Join(constants.StateDir, id+".yaml"), playbook, 0600); err != nil {
+		return nil, fmt.Errorf("cannot write playbook to temp file: %v", err)
+	}
+	if err := os.MkdirAll(filepath.Join(privateDataDir, "artifacts", id, "job_events"), 0755); err != nil {
+		return nil, fmt.Errorf("cannot create private state dir: %v", err)
+	}
+
+	// precreate the status file so that we can watch for when it gets written
+	// to.
+	status, err := os.Create(filepath.Join(privateDataDir, "artifacts", id, "status"))
+	if err != nil {
+		return nil, fmt.Errorf("cannot create status file: %v", err)
+	}
+	_ = status.Close()
+
+	events := make(chan Event)
+
+	// publish an "executor_on_start" event to signal cloud connector that a run
+	// event has started. This is run on a goroutine in case the events
+	// channel doesn't have a receiver connected yet to avoid blocking the
+	// continuation of this function.
+	go func() {
+		events <- Event{
+			Event:     "executor_on_start",
+			UUID:      uuid.New().String(),
+			Counter:   -1,
+			Stdout:    "",
+			StartLine: 0,
+			EndLine:   0,
+			EventData: struct {
+				CRCDispatcherCorrelationID string `json:"crc_dispatcher_correlation_id"`
+				CRCDispatcherErrorCode     string `json:"crc_dispatcher_error_code"`
+				CRCDispatcherErrorDetails  string `json:"crc_dispatcher_error_details"`
+			}{
+				CRCDispatcherCorrelationID: correlationID,
+			},
+		}
+	}()
+
+	// started is a closure that's passed to the StartProcess function as its
+	// exec.ProcessStartedFunc handler. This function is invoked on a goroutine
+	// after the process has been started. It exists as a named closure within
+	// this function to capture some relevant values that aren't otherwise
+	// passed in as part of the exec.ProcessStartedFunc signature (id,
+	// privateDataDir, etc.).
+	started := func(pid int, stdout, stderr io.ReadCloser) {
+		log.Debugf("process started: pid=%v runner_ident=%v", pid, id)
+
+		// start a goroutine to watch for event files being written to the
+		// job_events directory. When a relevant event file is detected, it gets
+		// marshaled into JSON and sent to the events channel.
+		go func() {
+			c := make(chan notify.EventInfo, 1)
+			if err := notify.Watch(filepath.Join(privateDataDir, "artifacts", id, "job_events"), c, notify.InMovedTo); err != nil {
+				log.Errorf("failed to watch private data dir: dir=%v err=%v", privateDataDir, err)
```

Comment:
This log message is not correct, because we actually watch `filepath.Join(privateDataDir, "artifacts", id, "job_events")`. This could be confusing.